google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionRunner.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.core.ApiAsyncFunction;
+import com.google.api.core.ApiFunction;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.CurrentMillisClock;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.retrying.ExponentialRetryAlgorithm;
+import com.google.api.gax.retrying.TimedAttemptSettings;
+import com.google.api.gax.rpc.ApiException;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.MoreExecutors;
+import io.opencensus.trace.AttributeValue;
+import io.opencensus.trace.Span;
+import io.opencensus.trace.Tracer;
+import io.opencensus.trace.Tracing;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implements backoff and retry semantics for Firestore transactions.
+ *
+ * <p>Consumers should invoke `run()` to obtain an ApiFuture that resolves when all retries
+ * complete. Backoff settings can be customized via FirebaseOptions.
+ */
+class TransactionRunner<T> {
+
+  private static final Tracer tracer = Tracing.getTracer();
+  private static final io.opencensus.trace.Status TOO_MANY_RETRIES_STATUS =
+      io.opencensus.trace.Status.ABORTED.withDescription("too many retries");
+  private static final io.opencensus.trace.Status USER_CALLBACK_FAILED =
+      io.opencensus.trace.Status.ABORTED.withDescription("user_callback_failed");
+
+  private final Transaction.AsyncFunction<T> userCallback;
+  private final Span span;
+  private final FirestoreImpl firestoreClient;
+  private final ScheduledExecutorService firestoreExecutor;
+  private final Executor userCallbackExecutor;
+  private final ExponentialRetryAlgorithm backoffAlgorithm;
+  private TimedAttemptSettings nextBackoffAttempt;
+  private Transaction transaction;
+  private int attemptsRemaining;
+
+  /**
+   * @param firestore The active Firestore instance
+   * @param userCallback The user provided transaction callback
+   * @param userCallbackExecutor The executor to run the user callback on
+   * @param numberOfAttempts The total number of attempts for this transaction
+   */
+  public TransactionRunner(
```

Comment:
This can be package-private.

```suggestion
  TransactionRunner(
```

Reply:
Fixed